--- old/vsock-only/linux/af_vsock.c	2016-01-25 14:35:20.000000000 +0300
+++ new/vsock-only/linux/af_vsock.c	2016-01-25 14:48:11.000000000 +0300
@@ -4642,7 +4642,7 @@
        * able to send.
        */
 
-      written = VMCIQPair_EnqueueV(vsk->qpair, msg->msg_iov,
+      written = VMCIQPair_EnqueueV(vsk->qpair, &msg->msg_iter.iov,
                                    len - totalWritten, 0);
       if (written < 0) {
          err = -ENOMEM;
@@ -4927,9 +4927,9 @@
    }
 
    if (flags & MSG_PEEK) {
-      copied = VMCIQPair_PeekV(vsk->qpair, msg->msg_iov, len, 0);
+      copied = VMCIQPair_PeekV(vsk->qpair, &msg->msg_iter.iov, len, 0);
    } else {
-      copied = VMCIQPair_DequeueV(vsk->qpair, msg->msg_iov, len, 0);
+      copied = VMCIQPair_DequeueV(vsk->qpair, &msg->msg_iter.iov, len, 0);
    }
 
    if (copied < 0) {
