--- old/vmnet-only/filter.c	2016-01-25 14:16:01.000000000 +0300
+++ new/vmnet-only/filter.c	2016-01-25 14:28:50.000000000 +0300
@@ -214,9 +214,13 @@
 #else
                  struct sk_buff **pskb,                // IN:
 #endif
+#if LINUX_VERSION_CODE < KERNEL_VERSION(4, 1, 15)
                  const struct net_device *in,          // IN:
                  const struct net_device *out,         // IN:
                  int (*okfn)(struct sk_buff *))        // IN:
+#else
+                 const struct nf_hook_state *state)
+#endif
 {
 #ifndef VMW_NFHOOK_USES_SKB
    struct sk_buff *skb = *pskb;
