--- old/vmnet-only/compat_netdevice.h	2016-01-25 14:13:20.000000000 +0300
+++ new/vmnet-only/compat_netdevice.h	2016-01-25 14:14:50.000000000 +0300
@@ -180,7 +180,11 @@
 }
 #   define compat_free_netdev(dev)     kfree(dev)
 #else
+#   if LINUX_VERSION_CODE < KERNEL_VERSION(3, 17, 0)
 #   define compat_alloc_netdev(size, mask, setup) alloc_netdev(size, mask, setup)
+#else
+#   define compat_alloc_netdev(size, mask, setup) alloc_netdev(size, mask, NET_NAME_USER, setup)
+#endif
 #   define compat_free_netdev(dev)                free_netdev(dev)
 #endif
 
