Description: Fix compilation failure on !linux
Author: Mattia Rizzolo <mattia@debian.org>
Bug: https://sourceforge.net/p/podofo/tickets/31
Last-Update: 2018-09-22

--- a/test/unit/ParserTest.cpp
+++ b/test/unit/ParserTest.cpp
@@ -2037,7 +2037,7 @@
 #elif defined( __linux__ )
     // TODO do big allocs succeed then trigger OOM-killer fiasco??
     bool bCanTerminateProcess = false;
-#elif
+#else
     // other systems - assume big allocs faily gracefully and throw bad_alloc
     bool bCanTerminateProcess = false;
 #endif
