--- a/libgitg/gitg-lanes.vala	2017-08-09 18:34:17.000000000 +0300
+++ b/libgitg/gitg-lanes.vala	2018-11-15 14:48:46.000000000 +0300
@@ -338,6 +338,8 @@
 			{
 				unowned Lane lane = lns.nth_data(index);
 
+			    if (lane != null)
+			    {
 				if (item.next != null)
 				{
 					var newindex = lane.from.data;
@@ -363,6 +365,7 @@
 					lane.tag |= LaneTag.END;
 					lane.boundary_id = container.to;
 				}
+			    }
 			}
 
 			item = item.next;
