--- old/opencascade-6.7.1/inc/OpenGl_GlCore11.hxx	2014-04-29 13:40:37.000000000 +0400
+++ new/opencascade-6.7.1/inc/OpenGl_GlCore11.hxx	2015-01-05 14:29:33.000000000 +0300
@@ -34,6 +34,7 @@
 
 // exclude modern definitions and system-provided glext.h, should be defined before gl.h inclusion
 #define GL_GLEXT_LEGACY
+#define GLX_GLXEXT_LEGACY
 
 // include main OpenGL header provided with system
 #if defined(__APPLE__)
